### PR TITLE
feat: do not ask for client secret when using auth code with PKCE

### LINKED ISSUE #6290: Authorization Flow w/PKCE asks for Client Secret
### Q&A (please complete the following information)
 - OS: Windows 10
 - Browser: Chrome
 - Version: 84
 - Method of installation: Swashbuckle
 - Swagger-UI version: 3.26
 - Swagger/OpenAPI version: Swagger 2.0, OpenAPI 3.0

### Content & configuration

Any API with Authorization Flow and PKCE enabled.

I'm using DotNet Core and Swashbuckle, so it's a little hard to provide anything here.

### Describe the bug you're encountering

The Authorize UI still prompts for the Client Secret even though it's not required for the PKCE flow.

### To reproduce...

1. Click on 'Authorize'
2. See 'client_secret' field is shown (empty)
  a. Passing any value (valid or not) in that field and clicking 'Authorize' will cause an error.
  b. Leaving the field blank, will cause the auth flow to work correctly.

### Expected behavior
When 'UsePKCE' flag is enabled, Client Secret field is not shown.

### Screenshots
image

### PR DESCRIPTION
**Pull request by ch-egli**
According to [OAuth 2.0 Security Best Current Practice](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-security-topics)   the implicit grant should no longer be used, but be replaced by the authorization code grant with PKCE. 
While the swagger-ui permits to use the PKCE grant type, the authorization dialog still prompts for a client secret. From a user perspective, this is confusing and the current pull request omits the client secret field if authorization code grant with PKCE is configured (see https://github.com/swagger-api/swagger-ui/issues/6290).

### Description
When the flag `usePkceWithAuthorizationCodeGrant` is set to true, the input field for the client secret will not be shown. 
Moreower, we specify in the authorization dialogue, that the authorization code flow with PKCE is used. 

### Motivation and Context
See description above. 
Fixes #6290 

### How Has This Been Tested?

- generated a new dist and included it in different projects of Schweizerische Bundesbahnen (SBB)
- cypress e2e test added

### Screenshots (if appropriate):
auth-code-grant-dialog
auth-code-grant-dialog-with-pkce

### My PR contains... 

- [ ] No code changes (`src/` is unmodified: changes to documentation, CI, metadata, etc.)
- [ ] Dependency changes (any modification to dependencies in `package.json`)
- [ ] Bug fixes (non-breaking change which fixes an issue)
- [ ] Improvements (misc. changes to existing features)
- [x] Features (non-breaking change which adds functionality)

### My changes...
- [ ] are breaking changes to a public API (config options, System API, major UI change, etc).
- [ ] are breaking changes to a private API (Redux, component props, utility functions, etc.).
- [ ] are breaking changes to a developer API (npm script behavior changes, new dev system dependencies, etc).
- [x] are not breaking changes.

### Documentation
- [x] My changes do not require a change to the project documentation.
- [ ] My changes require a change to the project documentation.
- [ ] If yes to above: I have updated the documentation accordingly.

### Automated tests
- [ ] My changes can not or do not need to be tested.
- [x] My changes can and should be tested by unit and/or integration tests.
- [x] If yes to above: I have added tests to cover my changes.
- [ ] If yes to above: I have taken care to cover edge cases in my tests.
- [x] All new and existing tests passed.

### CODE REVIEW

**Review by Messhias — APPROVED** [review]

**Comment by Messhias** [comment]
@swaggerhub-bot anyone to check this?

**Comment by ilozano2** [comment]
This looks really good @Messhias 

I'd appreciate having this improvement merged. Is there anything I can do to help?

**Comment by Messhias** [comment]
> This looks really good @Messhias
> 
> I'd appreciate having this improvement merged. Is there anything I can do to help?

You can update the current PR with the base branch for workflow approval.

**Comment by ilozano2** [comment]
@Messhias I don't have permission to update @ch-egli fork. Can I create a new PR picking his commits to get a more updated PR?

**Comment by Messhias** [comment]
> @Messhias I don't have permission to update @ch-egli fork. Can I create a new PR picking his commits to get a more updated PR?

In fact you should be able to do updates and push commits on this branch.

**Comment by ilozano2** [comment]
@Messhias I cannot edit any file on this PR (I guess I am not a maintainer on this repo or the PR wasn't marked to be editable by maintainers).

I've just created a copy of this PR https://github.com/swagger-api/swagger-ui/pull/7498 [enabling the edition from maintainers](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).

**Comment by Messhias** [comment]
> @Messhias I cannot edit any file on this PR (I guess I am not a maintainer on this repo or the PR wasn't marked to be editable by maintainers).
> 
> I've just created a copy of this PR #7498 [enabling the edition from maintainers](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).

Ok, I'll check this new PR and let you know. Thank you.

**Comment by ch-egli** [comment]
Hi all
I have marked the pull request with `Allow edits by maintainers`. 
Can I do anything else?

**Comment by Messhias** [comment]
> Hi all
> I have marked the pull request with `Allow edits by maintainers`.
> Can I do anything else?

No, only the repository owner/maintainers can do the stuff of merging and etc... What could be done is the repository owner add more maintainers to keep the package evolve more constant and quickly.

File: src/core/components/auth/oauth2.jsx [hunk 1]
```diff
@@ -130,7 +130,11 @@ export default class Oauth2 extends React.Component {
     const AUTH_FLOW_ACCESS_CODE = isOAS3() ? (oidcUrl ? "authorization_code" : "authorizationCode") : "accessCode"
     const AUTH_FLOW_APPLICATION = isOAS3() ? (oidcUrl ? "client_credentials" : "clientCredentials") : "application"
 
+    let authConfigs = authSelectors.getConfigs()
+    let isPkceCodeGrant = authConfigs.usePkceWithAuthorizationCodeGrant === "true" || authConfigs.usePkceWithAuthorizationCodeGrant === true
+
     let flow = schema.get("flow")
+    let flowToDisplay = isPkceCodeGrant ? flow + " with PKCE" : flow
```
Comment (ponelat) [inline src/core/components/auth/oauth2.jsx:137]: pkce only applies to authorization flow yeah? This would show " with PKCE" even if the implicit flow is used. Perhaps we should ignore the usePkCE flag if flows _aren't_ authorization.

**Review by ponelat — COMMENTED** [review]
Thanks @ch-egli and @Messhias @ilozano2 ! This is great work, first time I'm learning about the PKCE extension. Added a small comment. 

Will try to get this in the pipeline for a core contributor to review.

File: src/core/components/auth/oauth2.jsx (hunk 1 above)
Comment (ilozano2) [reply src/core/components/auth/oauth2.jsx:137]: AFAIK, PKCE is an improvement on top of the authorization_code so public clients don't need to expose the client_secret in their code. 

I would keep the pkce flag there because there is no way to determine if pkce is enabled in the server using the openid-configuration file. On the other hand, even a user can figure out that it is pkce because client_secret is not presented, I'd keep "with PKCE" title ending.

The code looks good to me

Comment (ch-egli) [reply src/core/components/auth/oauth2.jsx]: @ponelat: Thank you for pointing this out to me. You are correct, you really only want to show this additional text if authorizationCode flow is used. I have updated the corresponding code in my fork of the repository by adding an additional test.

**Review by ilozano2 — APPROVED** [review]

**Comment by ponelat** [comment]
@ilozano2 thanks for the ping. I think it looks just dandy, we may not get a merge this week, as we finish off the current release, but likely next week. cc @char0n

**Comment by ilozano2** [comment]
@ponelat any news about this improvement?

**Comment by char0n** [comment]
@ilozano2 yep, this is planned to be included in 4.3.0 release.

**Comment by ponelat** [comment]
Apologies @ilozano2 I dropped the ball on this one!

File: test/e2e-cypress/tests/features/auth-code-flow-pkce-without-secret.js [hunk 2]
```diff
@@ -0,0 +1,41 @@
+describe("Check client_secret for OAuth2 Authorization Code flow with and without PKCE (#6290)", () => {
+  it("should not display client_secret field for authorization code flow with PKCE", () => {
+    cy.visit(
+      "/?url=/documents/features/auth-code-flow-pkce-without-secret.yaml"
+    )
+      .window()
+      .then(win => {
+        // set auth config to use PKCE
+        let authConfigs = win.ui.authSelectors.getConfigs()
```
Comment (char0n) [inline test/e2e-cypress/tests/features/auth-code-flow-pkce-without-secret.js:9]: Could we avoid mutations here and use actions to allow SwaggerUI to properly react to the data change?

```suggestion
        let authConfigs = win.ui.authSelectors.getConfigs()
        win.ui.authActions.configureAuth({
          ...authConfigs,
          usePkceWithAuthorizationCodeGrant: true,
        })
```

File: test/e2e-cypress/tests/features/auth-code-flow-pkce-without-secret.js [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+  })
+
+  it("should display client_secret field for authorization code flow without PKCE", () => {
+    cy.visit(
+      "/?url=/documents/features/auth-code-flow-pkce-without-secret.yaml"
+    )
+      .window()
+      .then(win => {
+        // set auth config to not use PKCE
+        let authConfigs = win.ui.authSelectors.getConfigs()
```
Comment (char0n) [inline test/e2e-cypress/tests/features/auth-code-flow-pkce-without-secret.js:29]: Could we avoid mutations here and use actions to allow SwaggerUI to properly react to the data change?

```suggestion
        let authConfigs = win.ui.authSelectors.getConfigs()
        win.ui.authActions.configureAuth({
          ...authConfigs,
          usePkceWithAuthorizationCodeGrant: false,
        })
```

**Review by char0n — APPROVED** [review]
Thanks for contributing! This PR is good to go. There are two suggestions that I've made, which IMHO are more appropriate given SwaggerUI uses non-mutating mechanism of redux.

**Comment by chrklin** [comment]
Hmm does this break the client credentials. I have 2 flows, one with auth code that requires pkce and a second that requires client credentials. Seems like I can't add the secret for client_credentials flow anymore because I need pkce on my authorization code flow

**Comment by chrklin** [comment]
> Hmm does this break the client credentials. I have 2 flows, one with auth code that requires pkce and a second that requires client credentials. Seems like I can't add the secret for client_credentials flow anymore because I need pkce on my authorization code flow

`(flow === AUTH_FLOW_APPLICATION || flow === AUTH_FLOW_ACCESS_CODE || flow === AUTH_FLOW_PASSWORD) && !isPkceCodeGrant &&`

should have been

`(flow === AUTH_FLOW_APPLICATION || (flow === AUTH_FLOW_ACCESS_CODE && !isPkceCodeGrant) || flow === AUTH_FLOW_PASSWORD)  &&`
